core/src/main/java/org/elasticsearch/action/support/single/custom/TransportSingleCustomOperationAction.java

```diff
@@ -123,7 +124,32 @@ private AsyncSingleAction(Request request, ActionListener<Response> listener) {
             if (blockException != null) {
                 throw blockException;
             }
-            this.shardsIt = shards(clusterState, internalRequest);
+            ShardsIterator originalShardsIt = shards(clusterState, internalRequest);
```

Comment:
Before if `preferLocal` was true, it would always first try on the local node before trying remote nodes. I assume we want to keep this behaviour? 

Before this was implemented by looping over the shard iterator and skipping remote shard routing, if the operation didn't complete successfully on a local node, the shard iterator was reset and it looped again over all the shard routings but no skipping local shard routings.

I agree that it is nicer of this local shard routings first logic is folded into a shard iterator impl.